File: turms-client-swift/Sources/TurmsClient/Driver/Service/ConnectionService.swift
import Foundation

private class MessageDecoder {
    private static let maxReadBufferCapacity = 8 * 1024 * 1024

    private var readIndex = 0
    private var tempPayloadLength = 0
    private var payloadLength: Int?

    private var readBuffer = Data()

    func decodeMessages(_ data: Data) throws -> [Data] {
        if (readBuffer.count + data.count) > MessageDecoder.maxReadBufferCapacity {
            throw RuntimeError("The read buffer has exceeded the maximum size \(MessageDecoder.maxReadBufferCapacity)")
        }
        readBuffer.append(data)
        var messages: [Data] = []
        while true {
            let message = try tryReadMessage()
            guard let msg = message else {
                break
            }
            messages.append(msg)
        }
        return messages
    }

    func clear() {
        readIndex = 0
        tempPayloadLength = 0
        payloadLength = nil
        readBuffer.removeAll()
    }

    private func tryReadMessage() throws -> Data? {
        guard let length = try tryReadVarInt() else {
            return nil
        }
        payloadLength = length
        let end = readIndex + payloadLength!
        if readBuffer.count < end {
            return nil
        }
        let message = readBuffer.subdata(in: readIndex ..< end)
        readBuffer.removeSubrange(0 ..< end)
        readIndex = 0
        payloadLength = nil
        return message
    }

    private func tryReadVarInt() throws -> Int? {
        let length = readBuffer.count
        while readIndex < 5 {
            if readIndex >= length {
                return nil
            }
            let byte = readBuffer[readIndex]
            tempPayloadLength |= (Int(byte) & 0x7F) << (7 * readIndex)
            readIndex += 1
            if byte & 0x80 == 0 {
                let length = tempPayloadLength
                tempPayloadLength = 0
                return length
            }
        }
        throw RuntimeError("VarInt input too big")
    }
}

public class ConnectionService: BaseService {
    private let initialHost: String
    private let initialPort: UInt16
    private let initialConnectTimeout: TimeInterval

    private var disconnectContinuations: [UnsafeContinuation<Void, Never>] = []

    private var onConnectedListeners: [() -> Void] = []
    private var onDisconnectedListeners: [(Error?) -> Void] = []
    private var messageListeners: [(Data) -> Void] = []

    private let lock = Lock()
    private let decoder = MessageDecoder()

    init(stateStore: StateStore, host: String? = nil, port: UInt16? = nil, connectTimeout: TimeInterval? = nil) {
        initialHost = host ?? "127.0.0.1"
        initialPort = port ?? 11510
        initialConnectTimeout = connectTimeout ?? 30
        super.init(stateStore)
    }

    private func resetStates() {
        fulfillDisconnectContinuations()
    }

    // Listeners

    func addOnConnectedListener(_ listener: @escaping () -> Void) {
        onConnectedListeners.append(listener)
    }

    func addOnDisconnectedListener(_ listener: @escaping (Error?) -> Void) {
        onDisconnectedListeners.append(listener)
    }

    func addMessageListener(_ listener: @escaping (Data) -> Void) {
        messageListeners.append(listener)
    }

    private func notifyOnConnectedListeners() {
        for onConnectedListener in onConnectedListeners {
            onConnectedListener()
        }
    }

    private func notifyOnDisconnectedListeners(_ error: Error?) {
        for onDisconnectedListener in onDisconnectedListeners {
            onDisconnectedListener(error)
        }
    }

    private func notifyMessageListeners(_ message: Data) {
        for messageListener in messageListeners {
            messageListener(message)
        }
    }

    private func fulfillDisconnectContinuations() {
        lock.locked {
            repeat {
                disconnectContinuations.popLast()?.resume()
            } while !disconnectContinuations.isEmpty
        }
    }

    // Connection

    public func connect(host: String? = nil, port: UInt16? = nil, connectTimeout _: TimeInterval? = nil, useTls: Bool? = false, certificatePinning: CertificatePinning? = nil) async throws {
        if stateStore.isConnected {
            throw ResponseError(code: .clientSessionAlreadyEstablished)
        }
        resetStates()
        let tcp = TcpClient(onClosed: { [weak self] error in
            self?.onSocketClosed(error)
        }, onDataReceived: { [weak self] data in
            guard let s = self else { return }
            let messages = try s.decoder.decodeMessages(data)
            for message in messages {
                s.notifyMessageListeners(message)
            }
        })
        try await tcp.connect(host: host ?? initialHost, port: port ?? initialPort, useTls: useTls ?? false, certificatePinning: certificatePinning)
        onSocketOpened()
        stateStore.tcp = tcp
    }

    public func disconnect() async {
        await withUnsafeContinuation { continuation in
            if !stateStore.isConnected {
                return continuation.resume()
            }
            lock.locked {
                disconnectContinuations.append(continuation)
            }
            stateStore.tcp!.close()
        }
    }

    // Lifecycle hooks

    private func onSocketOpened() {
        stateStore.isConnected = true
        notifyOnConnectedListeners()
    }

    private func onSocketClosed(_ error: Error?) {
        decoder.clear()
        stateStore.isConnected = false
        fulfillDisconnectContinuations()
        notifyOnDisconnectedListeners(error)
    }

    // Base methods

    override func close() async {
        return await disconnect()
    }
}

File: turms-client-dart/lib/src/driver/service/connection_service.dart
import 'dart:async';
import 'dart:io';

import 'package:collection/collection.dart';

import '../../../turms_client.dart';

typedef OnConnectedListener = void Function();
typedef OnDisconnectedListener = void Function(
    {Object? error, StackTrace? stackTrace});
typedef MessageListener = void Function(List<int> message);

class _MessageDecoder {
  static const int _maxReadBufferCapacity = 8 * 1024 * 1024;

  int _readIndex = 0;
  int _tempPayloadLength = 0;
  int? _payloadLength;

  final List<int> _readBuffer = [];

  final void Function(List<int> message) onDecoded;

  _MessageDecoder(this.onDecoded);

  void decodeMessages(List<int> bytes) {
    if ((_readBuffer.length + bytes.length) > _maxReadBufferCapacity) {
      throw Exception(
          'The read buffer has exceeded the maximum size $_maxReadBufferCapacity');
    }
    _readBuffer.addAll(bytes);
    while (_tryReadMessage()) {}
  }

  void clear() {
    _readIndex = 0;
    _tempPayloadLength = 0;
    _payloadLength = null;
    _readBuffer.clear();
  }

  bool _tryReadMessage() {
    _payloadLength ??= _tryReadVarInt();
    if (_payloadLength == null) {
      return false;
    }
    final end = _readIndex + _payloadLength!;
    if (_readBuffer.length < end) {
      return false;
    }
    final message = ListSlice(_readBuffer, _readIndex, end);
    onDecoded(message);
    _readBuffer.removeRange(0, end);
    _readIndex = 0;
    _payloadLength = null;
    return true;
  }

  int? _tryReadVarInt() {
    final length = _readBuffer.length;
    while (_readIndex < 5) {
      if (_readIndex >= length) {
        return null;
      }
      final byte = _readBuffer[_readIndex];
      _tempPayloadLength |= (byte & 0x7F) << (7 * _readIndex);
      _readIndex++;
      if (byte & 0x80 == 0) {
        final length = _tempPayloadLength;
        _tempPayloadLength = 0;
        return length;
      }
    }
    throw Exception('VarInt input too big');
  }
}

class ConnectionService extends BaseService {
  final String _initialHost;
  final int _initialPort;
  final int _initialConnectTimeoutMillis;

  final List<OnConnectedListener> _onConnectedListeners = [];
  final List<OnDisconnectedListener> _onDisconnectedListeners = [];
  final List<MessageListener> _messageListeners = [];

  late _MessageDecoder _decoder;

  factory ConnectionService(StateStore stateStore, String? host, int? port,
      int? connectTimeoutMillis) {
    final connectionService =
        ConnectionService._(stateStore, host, port, connectTimeoutMillis);
    connectionService._decoder =
        _MessageDecoder(connectionService._notifyMessageListeners);
    return connectionService;
  }

  ConnectionService._(
      super.stateStore, String? host, int? port, int? connectTimeoutMillis)
      : _initialHost = host ?? '127.0.0.1',
        _initialPort = port ?? 11510,
        _initialConnectTimeoutMillis = connectTimeoutMillis ?? 30 * 1000;

  // Listeners

  void addOnConnectedListener(OnConnectedListener listener) =>
      _onConnectedListeners.add(listener);

  void addOnDisconnectedListener(OnDisconnectedListener listener) =>
      _onDisconnectedListeners.add(listener);

  void addMessageListener(MessageListener listener) =>
      _messageListeners.add(listener);

  void removeOnConnectedListener(OnConnectedListener listener) =>
      _onConnectedListeners.remove(listener);

  void removeOnDisconnectedListener(OnDisconnectedListener listener) =>
      _onDisconnectedListeners.remove(listener);

  void removeMessageListener(MessageListener listener) =>
      _messageListeners.remove(listener);

  void _notifyOnConnectedListeners() {
    for (final listener in _onConnectedListeners) {
      listener.call();
    }
  }

  void _notifyOnDisconnectedListeners(Object? error, StackTrace? stackTrace) {
    for (final listener in _onDisconnectedListeners) {
      listener.call(error: error, stackTrace: stackTrace);
    }
  }

  void _notifyMessageListeners(List<int> message) {
    for (final listener in _messageListeners) {
      listener.call(message);
    }
  }

  // Connection

  Future<void> connect(
      {String? host,
      int? port,
      int? connectTimeoutMillis,
      bool? useTls = false,
      SecurityContext? context}) async {
    if (stateStore.isConnected) {
      if (host == stateStore.tcp?.host && port == stateStore.tcp?.port) {
        return;
      } else {
        throw ResponseException(
            code: ResponseStatusCode.clientSessionAlreadyEstablished);
      }
    }
    final tcp = TcpClient(_onSocketClosed, _decoder.decodeMessages);
    connectTimeoutMillis ??= _initialConnectTimeoutMillis;
    final timeout = connectTimeoutMillis > 0
        ? Duration(milliseconds: connectTimeoutMillis)
        : null;
    await tcp.connect(host ?? _initialHost, port ?? _initialPort,
        useTls ?? false, context, timeout);
    stateStore.tcp = tcp;
    _onSocketOpened();
  }

  Future<void> disconnect() async {
    if (stateStore.isConnected) {
      stateStore.isConnected = false;
      await stateStore.tcp?.close();
    }
    _decoder.clear();
  }

  // Lifecycle hooks

  void _onSocketOpened() {
    stateStore.isConnected = true;
    _notifyOnConnectedListeners();
  }

  void _onSocketClosed({Object? error, StackTrace? stackTrace}) {
    _decoder.clear();
    stateStore.isConnected = false;
    _notifyOnDisconnectedListeners(error, stackTrace);
  }

  @override
  Future<void> close() => disconnect();

  @override
  void onDisconnected({Object? error, StackTrace? stackTrace}) {}
}
